helper/cert/source_gen.go

```diff
@@ -138,15 +139,14 @@ func (s *GenSource) Certificate(ctx context.Context, last *Bundle) (Bundle, erro
 
 	// Generate cert, set it on the result, and return
 	cert, key, err := s.generateCert()
-	if err == nil {
-		result.Cert = []byte(cert)
-		result.Key = []byte(key)
+	if err != nil {
+		return result, err
+	}
+	result.Cert = []byte(cert)
+	result.Key = []byte(key)
 
-		if s.LeaderElector != nil {
-			if err := s.updateSecret(ctx, result); err != nil {
-				return result, fmt.Errorf("failed to update Secret: %s", err)
-			}
-		}
+	if s.LeaderElector != nil {
+		s.retryUpdateSecret(ctx, leaderCh, result)
```

Comment:
This sync call might affect `<-retryTicker.C` in `Notify.Run()`? For instance, if this call blocks for 6s it will be skipping the next tick (and every increment of 5 seconds that this blocks skips 1 tick). This may or may not be fine depending if there are other things within that loop that is expecting the ticker to run every 5 seconds.

Reply:
This looks fine to me, the `Notify.Run()` loop only notifies upstream when there is a new result to report, so I think it should be ok to reattempt until we get something worth reporting, and we're still cancellable via the context.